cpp/src/parquet/level_conversion_test.cc

```diff
@@ -108,22 +114,245 @@ TEST(GreaterThanBitmap, GeneratesExpectedBitmasks) {
+
+  std::vector<typename TypeParam::ListLengthType> lengths(5, 0);
+  uint64_t validity_output;
+  ValidityBitmapInputOutput validity_io;
+  validity_io.values_read_upper_bound = 4;
+  validity_io.valid_bits = reinterpret_cast<uint8_t*>(&validity_output);
+  typename TypeParam::ListLengthType* next_position = this->converter_.ComputeListInfo(
+      this->test_data_, level_info, &validity_io, lengths.data());
+
+  EXPECT_THAT(next_position, lengths.data() + 4);
+  EXPECT_THAT(lengths, testing::ElementsAre(0, 3, 7, 7, 7));
+
+  EXPECT_EQ(validity_io.values_read, 4);
+  EXPECT_EQ(validity_io.null_count, 1);
+  EXPECT_EQ(BitmapToString(validity_io.valid_bits, /*length=*/4), "1101");
+}
+
+TYPED_TEST(NestedListTest, MiddleListTest) {
+  // [null, [[1 , null, 3], []], []],
+  // [[[]], [[], [1, 2]], null, [[3]]],
+  // null,
+  // []
+  // -> middle lists (null, len(2), len(0),
+  //                  len(1), len(2), null, len(1),
+  //                  N/A,
+  //                  N/A
+  LevelInfo level_info;
+  level_info.rep_level = 2;
+  level_info.def_level = 4;
+  level_info.repeated_ancestor_def_level = 2;
+
+  std::vector<typename TypeParam::ListLengthType> lengths(8, 0);
+  uint64_t validity_output;
+  ValidityBitmapInputOutput validity_io;
+  validity_io.values_read_upper_bound = 7;
+  validity_io.valid_bits = reinterpret_cast<uint8_t*>(&validity_output);
+  typename TypeParam::ListLengthType* next_position = this->converter_.ComputeListInfo(
+      this->test_data_, level_info, &validity_io, lengths.data());
+
+  EXPECT_THAT(next_position, lengths.data() + 7);
+  EXPECT_THAT(lengths, testing::ElementsAre(0, 0, 2, 2, 3, 5, 5, 6));
+
+  EXPECT_EQ(validity_io.values_read, 7);
+  EXPECT_EQ(validity_io.null_count, 2);
+  EXPECT_EQ(BitmapToString(validity_io.valid_bits, /*length=*/7), "0111101");
+}
+
+TYPED_TEST(NestedListTest, InnerMostListTest) {
+  // [null, [[1, null, 3], []], []],
+  // [[[]], [[], [1, 2]], null, [[3]]],
+  // null,
+  // []
+  // -> 4 inner lists (N/A, [len(3), len(0)], N/A
+  //                        len(0), [len(0), len(2)], N/A, len(1),
+  //                        N/A,
+  //                        N/A
+  LevelInfo level_info;
+  level_info.rep_level = 3;
+  level_info.def_level = 6;
+  level_info.repeated_ancestor_def_level = 4;
+
+  std::vector<typename TypeParam::ListLengthType> lengths(7, 0);
+  uint64_t validity_output;
+  ValidityBitmapInputOutput validity_io;
+  validity_io.values_read_upper_bound = 6;
+  validity_io.valid_bits = reinterpret_cast<uint8_t*>(&validity_output);
+  typename TypeParam::ListLengthType* next_position = this->converter_.ComputeListInfo(
+      this->test_data_, level_info, &validity_io, lengths.data());
+
+  EXPECT_THAT(next_position, lengths.data() + 6);
+  EXPECT_THAT(lengths, testing::ElementsAre(0, 3, 3, 3, 3, 5, 6));
+
+  EXPECT_EQ(validity_io.values_read, 6);
+  EXPECT_EQ(validity_io.null_count, 0);
+  EXPECT_EQ(BitmapToString(validity_io.valid_bits, /*length=*/6), "111111");
+}
```

Comment:
Not sure what you think about this?

Reply:
sorry, I didn't respond directly here.  I'll refactor the tests tonight.

Reply:
For this, I think if we have good coverage at the reconstruction level it is likely better.  I refactored the tests so if we do identify bugs in the future it will be easy to add.   

My main motivation here is I missed a couple of bugs by focusing heavily on the analogous code and not end-to-end (this was partially hampered by not having the read side done).